Frameworks/ImageIO/CGImageDestination.mm

```diff
@@ -28,75 +885,358 @@ CGImageDestinationRef CGImageDestinationCreateWithDataConsumer(CGDataConsumerRef
+ @Notes The CFTypeID for an opaque type changes from release to release and so has been hard-coded in current implementation
 */
 CFTypeID CGImageDestinationGetTypeID() {
-    UNIMPLEMENTED();
-    return StubReturn();
+    CFTypeID imageTypeID = 269;
+    return imageTypeID;
 }
 
 /**
- @Status Stub
- @Notes
+ @Status Caveat
+ @Notes Most image properties are actually handled as part of AddImage, and SetProperties should only handle properties
+        that are applied to all frames of an image. As such, currently only GIF loop count is set here.
 */
 void CGImageDestinationSetProperties(CGImageDestinationRef idst, CFDictionaryRef properties) {
-    UNIMPLEMENTED();
+    if (!idst) {
+        return;
+    }
+
+    ImageDestination* imageDestination = (ImageDestination*)idst;
+
+    // If Encoder or is missing, the destination has either been already finalized or not initialized yet, so return
+    if (!imageDestination.idEncoder) {
+        NSTraceInfo(TAG, @"CGImageDestinationFinalize did not find an Encoder\n");
+        return;
+    }
+
+    // Looping properties for GIFs
+    if (imageDestination.type == typeGIF) {
+        char loopCountMSB = 0;
+        char loopCountLSB = 0;
+
+        if (properties && CFDictionaryContainsKey(properties, kCGImagePropertyGIFDictionary)) {
+            CFDictionaryRef gifDictionary = (CFDictionaryRef)CFDictionaryGetValue(properties, kCGImagePropertyGIFDictionary);
+            if (CFDictionaryContainsKey(gifDictionary, kCGImagePropertyGIFLoopCount)) {
+                int loopCount = [(id)CFDictionaryGetValue(gifDictionary, kCGImagePropertyGIFLoopCount) intValue];
+                loopCountLSB = loopCount & 0xff;
+                loopCountMSB = (loopCount >> 8) & 0xff;
+            }
+        }
+
+        ComPtr<IWICMetadataQueryWriter> imageMetadataQueryWriter = imageDestination.idGifEncoderMetadataQueryWriter;
+        PROPVARIANT propertyToWrite;
+
+        PropVariantInit(&propertyToWrite);
+        propertyToWrite.vt = VT_VECTOR | VT_UI1;
+        propertyToWrite.caub.cElems = 11;
+        propertyToWrite.caub.pElems = (unsigned char*)[@"NETSCAPE2.0" UTF8String];
+        writePropertyToFrame(&propertyToWrite, L"/appext/Application", imageMetadataQueryWriter.Get());
+
+        propertyToWrite.caub.cElems = 5;
+        propertyToWrite.caub.pElems =
+            (unsigned char*)[[NSString stringWithFormat:@"%c%c%c%c%c", 3, 1, loopCountLSB, loopCountMSB, 0] UTF8String];
+        writePropertyToFrame(&propertyToWrite, L"/appext/Data", imageMetadataQueryWriter.Get());
+    }
 }
 
 /**
- @Status Stub
- @Notes
+ @Status Caveat
+ @Notes Current release supports JPEG, BMP, PNG, GIF, & TIFF image formats only 
 */
 bool CGImageDestinationFinalize(CGImageDestinationRef idst) {
-    UNIMPLEMENTED();
-    return StubReturn();
-}
+    if (!idst) {
+        return false;
+    }
+
+    ImageDestination* imageDestination = (ImageDestination*)idst;
+
+    if (imageDestination.count != imageDestination.maxCount) {
```

Comment:
thanks for the due diligence here

---
In reply to: [60517955](https://github.com/Microsoft/WinObjC/pull/445#discussion_r60517955) [](ancestors = 60517955)